### PR TITLE
Add external credit support

### PR DESCRIPTION
**Pull request by bnehoran**
- Implement external credit support in a separate tab in the main view
- Use styled components instead of a css file to style the schedule + semesters
- Refactor the schedule by using css grid instead of tables, which allows for more flexibility
- Update color for courses that can fit in both semesters to a gradient
- Fix bug where the delete course button was overflowing to the next line in small widths
- Update tutorial colors

Resolves https://github.com/TigerPathApp/tigerpath/issues/101
Resolves https://github.com/TigerPathApp/tigerpath/issues/179

Schedule view:

image

External Credits view:

image

### CODE REVIEW

**Comment by bnehoran** [comment]
This looks really good!

The only issue I can see is that when you click on an external credit in the req tree, it disappears from the req tree and there's no way to bring it back other than to delete the class and readd it.

Also, we've discussed this, and maybe it doesn't have to be part of the MVP, but should be added at some point: there should be a way to pin additional requirements on external courses that have already been added. This is both for external credits that satisfy multiple requirements, and so that the user doesn't have to delete and readd them when they switch majors.

File: frontend/src/components/App.js [hunk 1]
```diff
@@ -61,7 +53,12 @@ export default class App extends Component {
     for (let semIndex = 0; semIndex < schedule.length; semIndex++) {
       strippedSchedule.push([]);
       for (let course of schedule[semIndex]) {
-        strippedSchedule[semIndex].push({id: course['id'], settled: course['settled']});
+        let strippedCourse = { id: course['id'], settled: course['settled'] };
```
Comment (bnehoran) [inline frontend/src/components/App.js:56]: How come there's no name for non-external courses?

File: frontend/src/components/ExternalCreditForm.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                        <Dropdown.Item key={semName} eventKey={index} onSelect={(e) => this.handleChange("selectedSemester", semName, e)}>
+                          {semName}
+                        </Dropdown.Item>
+                      )
+                    }
+                  </Dropdown.Menu>
+                </SemesterDropdown>
+                <Form.Text className="text-muted">
+                  For AP credits, waivers, or other external credits that don't fit in a semester, choose "N/A" - they will appear in "Your External Credits" on the left.
+                </Form.Text>
```
Comment (bnehoran) [inline frontend/src/components/ExternalCreditForm.js]:
```suggestion
                  For summer courses, choose either the semester before or after in which you want it to appear.
                </Form.Text>
```

File: frontend/src/components/ExternalCreditForm.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                      getSemesterNames(profile.classYear).map((semName, index) =>
+                        <Dropdown.Item key={semName} eventKey={index} onSelect={(e) => this.handleChange("selectedSemester", semName, e)}>
+                          {semName}
+                        </Dropdown.Item>
+                      )
+                    }
+                  </Dropdown.Menu>
+                </SemesterDropdown>
+                <Form.Text className="text-muted">
+                  For AP credits, waivers, or other external credits that don't fit in a semester, choose "N/A" - they will appear in "Your External Credits" on the left.
```
Comment (bnehoran) [inline frontend/src/components/ExternalCreditForm.js]: What's going on here? Is the apostrophe / single quote supposed to be escaped?

File: frontend/src/styles/Courses.css [hunk 4]
```diff
@@ -42,8 +42,8 @@
 }
 
 .search-card.both {
-  background-color: #fbf2e9;
-  color: #753a00;
+  background: linear-gradient(to right, #f6ebff, #ebf6f9);
```
Comment (bnehoran) [inline frontend/src/styles/Courses.css:45]: This looks a bit odd, but I guess I could get used to it.
It will probably make the color scheme more intuitive, so I think it's an improvement.

File: tigerpath/static/app/js/tutorial.js [hunk 5]
```diff
@@ -11,7 +11,7 @@ function startIntro() {
         intro: `You can search for courses you want to add to your schedule here. This can be by the course code or title.
         Try searching something!
```
Comment (bnehoran) [inline tigerpath/static/app/js/tutorial.js]: It seems that with the newer rewording, I've observed that users no longer take it to mean literally do it now.
This is not really related to this PR, bit it's a small enough change that it probably doesn't merit its own.
```suggestion
         Try searching something now!
```

(and since you're already making changes to this spot in this PR)

File: tigerpath/views.py [hunk 6]
```diff
@@ -242,6 +246,11 @@ def update_schedule(request):
 def get_schedule(request):
     curr_user = request.user.profile
     schedule = populate_user_schedule(curr_user.user_schedule)
+
+    # make sure that the schedule has 9 semesters
+    if len(schedule) < 9:
```
Comment (bnehoran) [inline tigerpath/views.py]:
```suggestion
    while len(schedule) < 9:
```
I don't know if this is needed right now, since all users will presently have either 8 or 9, but it would cover any future changes in schedule length.

Comment (bnehoran) [reply tigerpath/views.py]: Like, suppose we expand to 10 at some point, and some user still has 8.

**Comment by danielleung27** [comment]
I agree with Barak that clicking on a req on the tree from external credit and making it go away doesn't really make sense (unless that's a bug) since the course is still on the schedule. Other than that, everything looks really good!

**Review by danielleung27 — DISMISSED** [review]
I agree with Barak that clicking on a req on the tree from external credit and making it go away doesn't really make sense (unless that's a bug) since the course is still on the schedule. Other than that, everything looks really good! I think the bug is small enough that we can still just launch right now.

File: frontend/src/components/App.js (hunk 1 above)
Comment (churichard) [reply frontend/src/components/App.js:56]: The name is retrieved from the database.

File: frontend/src/components/ExternalCreditForm.js (hunk 3 above)
Comment (churichard) [reply frontend/src/components/ExternalCreditForm.js]: I think this is just a bug in GitHub's syntax highlighting. The code runs fine and there are no errors.

File: frontend/src/styles/Courses.css (hunk 4 above)
Comment (churichard) [reply frontend/src/styles/Courses.css:45]: I think it's an improvement over before as well, but I mainly just needed to change the color to something else in order to accommodate external credits. I'll keep thinking about ways to improve our color scheme (for example, now external credits stand out in the schedule more than I'd like them to).

**Comment by churichard** [comment]
Clicking on the external credit in the requirements pane and having it go away is definitely a bug. Good catch, guys! I've fixed it and also addressed Barak's comments.

**Review by bnehoran — COMMENTED** [review]
Looks good!

**Review by churichard — APPROVED** [review]
